### DISCUSSION TITLE
How pointers are meant to be used

### DISCUSSION THREAD

**Discussion by amukh1** [cc65/cc65#2427] [Q&A] [answered]
When I make a pointer in C (cc65 obviously), it automatically allocates two bytes for it, since the address bus on the 6502 is 16 bits long which makes sense. But when I go to use these 16-bit pointers, the high byte gets chopped off and it is stuck pointing somewhere on the zero-page. So my question is, are pointers only _supposed_ to point to things ONLY on the zero page (because of the design of the 6502)? Or is this something I need to fix on my setup. If this is something I need to fix, how would I go about doing that?

I think I should mention if I explicitly hard-code the 16-bit pointer, which makes sense and would probably correspond to the 0xAD and 0x8D instructions, it isn't so clear how cc65 would even handle 16 bit pointers on the 6502, since the indirect operations are only on the zero-page.

Here is an example:

```c
char* y;

int main() {
    *(char*)0x00 = 0x01; // works
    
    y = (char*)0x1001;
    *y = 0x02; // doesnt work, or well does but allocates to 0x0001/0x01 instead of 0x1001
    
    return 0;
}

```

**Comment by mrdudz** [comment, marked as answer]
Seems to produce the expected result for me

```
$ cl65 --add-source -O -S -o test.s test.c

;
; *(char*)0x00 = 0x01; // works
;
        lda     #$01
        sta     $0000
;
; y = (char*)0x1001;
;
        ldx     #$10
        sta     _y
        stx     _y+1
;
; *y = 0x02; // doesnt work, or well does but allocates to 0x0001/0x01 instead of 0x1001
;
        sta     ptr1
        stx     ptr1+1
        lda     #$02
        ldy     #$00
        sta     (ptr1),y
;
; return 0;
;
        ldx     #$00
        txa
;
; }
;
        rts
```

So no, pointers are not limited to zeropage, that would be really weird and not useful at all :)

**Reply by amukh1** [reply]
Hmm. Ykw maybe it is my emulator. I'm tripping 🤣 whats the use of zp restricted pointers. thanks
